### PR TITLE
feat: add checksum for checkpoint data 

### LINKED ISSUE #2164: implement checkpoint checksum
### What type of enhancement is this?

Other

### What does the enhancement do?

Manifest checkpoint has left a data slot for its checksum. Implementing it can make the system a little bit more robust

### Implementation challenges

_No response_

### PR DESCRIPTION
**Pull request by tisonkun**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

This closes #2164.

This supersedes https://github.com/GreptimeTeam/greptimedb/pull/3090.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

File: src/mito2/src/manifest/storage.rs [hunk 1]
```diff
@@ -380,7 +404,7 @@ impl ManifestObjectStore {
         let checkpoint_metadata = CheckpointMetadata {
             size: bytes.len(),
             version,
-            checksum: None,
+            checksum: Some(checksum),
```
Comment (tisonkun) [inline src/mito2/src/manifest/storage.rs:407]: I'm wondering if we can use `u32` here. Previous `null` can be decoded to default `0` and is always mismatched?

File: src/mito2/src/error.rs [hunk 2]
```diff
@@ -575,6 +575,9 @@ pub enum Error {
 
     #[snafu(display("Invalid region options, {}", reason))]
     InvalidRegionOptions { reason: String, location: Location },
+
+    #[snafu(display("checksum mismatch (got: {}, wanted: {})", got, wanted))]
```
Comment (killme2008) [inline src/mito2/src/error.rs:579]: Prefer `actual` and `expected` personally.
```suggestion
    #[snafu(display("checksum mismatch (actual: {}, expected: {})", got, wanted))]
```

File: src/mito2/src/manifest/storage.rs [hunk 3]
```diff
@@ -74,6 +75,28 @@ pub fn gen_path(path: &str, file: &str, compress_type: CompressionType) -> Strin
     }
 }
 
+#[inline]
```
Comment (killme2008) [inline src/mito2/src/manifest/storage.rs:78]: We don't use `#[inline]` anymore, leave it to compiler.
```suggestion
```

File: src/mito2/src/manifest/storage.rs [hunk 4]
```diff
@@ -74,6 +75,28 @@ pub fn gen_path(path: &str, file: &str, compress_type: CompressionType) -> Strin
...
+fn checkpoint_checksum(data: &[u8]) -> u32 {
+    let mut hasher = Hasher::new();
+    hasher.update(data);
+    hasher.finalize()
+}
+
+#[inline]
```
Comment (killme2008) [inline src/mito2/src/manifest/storage.rs:85]:
```suggestion
```

File: src/mito2/src/manifest/storage.rs (hunk 1 above)
Comment (killme2008) [reply src/mito2/src/manifest/storage.rs:407]: If so, it will break the compatibility?

File: src/mito2/src/manifest/storage.rs [hunk 5]
```diff
@@ -544,6 +619,15 @@ mod tests {
         ManifestObjectStore::new("/", object_store, CompressionType::Uncompressed)
     }
 
+    fn new_checkpoint_metadata_with_version(version: ManifestVersion) -> CheckpointMetadata {
```
Comment (killme2008) [inline src/mito2/src/manifest/storage.rs:622]: Looks like it's only used in test.
```suggestion
    #[cfg(test)]
    fn new_checkpoint_metadata_with_version(version: ManifestVersion) -> CheckpointMetadata {
```

**Comment by killme2008** [comment]
@evenyag PTAL

File: src/mito2/src/manifest/storage.rs (hunk 1 above)
Comment (tisonkun) [reply src/mito2/src/manifest/storage.rs:407]: You're right that we should then skip checksum verification on `0`, which is effectively the same as skip for `None`.

Keep the current impl.

**Comment by tisonkun** [comment]
@killme2008 all comments resolved.

**Review by killme2008 — APPROVED** [review]
LGTM 👍

**Pull request merged**